File: cpp/day07.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <regex>
#include <map>
#include <vector>
typedef std::vector<std::string> rule;
typedef std::map<std::string, rule> ruleMap;


using namespace std;

rule tokenize(string s, string del = " ") {
    int start = 0;
    int end = s.find(del);
    rule rules;
    while (end != -1) {
        rules.push_back(s.substr(start, end - start));
        start = end + del.size();
        end = s.find(del, start);
    }
    rules.push_back(s.substr(start, end - start));
    return(rules);
}

ruleMap convert_line(string row){
    string toErase = " bags";
    string stopwords [2] = {" bags", " bag"};
    for (string word: stopwords){
        regex pattern(word);
        row = regex_replace(row, pattern, "");
    }
    row = row.substr(0, row.size()-1);

    string delimiter = " contain ";
    string outer_color = row.substr(0, row.find(delimiter));
    string inner = row.substr(row.find(delimiter) + delimiter.length());
    rule inner_str = tokenize(inner, ", ");
    rule inner_colors;
    for (string desc: inner_str)
    {
        string inner = desc.substr(desc.find(" ") + 1);
        inner_colors.push_back(inner);
    }
    if (find(inner_colors.begin(), inner_colors.end(), "other") != inner_colors.end())
    {
        while (!inner_colors.empty()) {
            inner_colors.pop_back();
        }
    }
    ruleMap m;
    m[outer_color] = inner_colors;
    return(m);
}

int has_gold(rule colors, ruleMap rules){
    if (colors.size() == 0){
        return(0);
    }
    else {
        string color = colors[0];
        if (color == "shiny gold"){
            return(1);
        }
        else{
            colors.erase(colors.begin());
            for (string c: rules[color]){
                colors.push_back(c);
            }
            return(has_gold(colors, rules));
        }
    }
}

int main () {
    string line;
    ifstream myfile ("../data/day07.txt");
    if (myfile.is_open()) 
    {
        ruleMap rules;
        while ( getline (myfile, line) ) 
        {
            ruleMap m = convert_line(line);
            rules.insert(m.begin(), m.end());
        }
        cout << "finished reading \n";
        int count = 0;
        for (auto const& [key, val] : rules)
        {
            count += has_gold(rules[key], rules);
        }
        myfile.close();
        cout << count << "\n";
    }
    else cout << "Unable to open file"; 
}

File: python/day07.py
import sys
sys.setrecursionlimit(1500)

with open('../data/day07.txt', 'r') as f:
    dat = [line.replace('\n', '') for line in f]


def convert_line(line):
    """ parse string and format into dict
    """
    line = line.replace(' bags', '')
    line = line.replace(' bag', '')
    line = line.replace('.', '')

    outer_color, inner = line.split(' contain ')
    inner_str = inner.split(', ')
    inner = [desc.split(' ', 1) for desc in inner_str]
    inner_colors = [desc[1] for desc in inner]

    if 'other' in inner_colors:
        inner_colors = []
    return outer_color, inner_colors


def has_gold(colors, rules_dict):
    """ recursively check all the colors and their colors
    """
    if len(colors) == 0:
        return 0
    else:
        color = colors[0]
        if color == 'shiny gold':
            return 1
        else:
            # add inner colors to queue
            return has_gold(rules_dict[color] + colors[1:], rules_dict)


# convert raw data into dictionary containing list of colors 
rules_dict = {}
for line in dat:
    line = convert_line(line)
    rules_dict[line[0]] = line[1]

results = [has_gold(colors, rules_dict) for colors in rules_dict.values()]
print(sum(results))


## Part II

def convert_line(line):
    
    line = line.replace(' bags', '')
    line = line.replace(' bag', '')
    line = line.replace('.', '')
    
    outer, inner = line.split(' contain ')
    inner = inner.split(', ')
    inner = [desc.split(' ', 1) for desc in inner]
        
    d = [(key, int(value)) for value, key in inner if key != 'other']

    return outer, d


def count_bags(colors):
    
    if len(colors) == 0:
        return 0
    else:
        color_pair = colors[0]
        color = color_pair[0]
        bags = color_pair[1]
        
        # sum the number of bags, the bags inside of the bags, and the rest of the uncounted bags
        return bags + bags * count_bags(datdict[color]) + count_bags(colors[1:])

# convert raw data into dictionary containing list of color-number pairs
datdict = {}
for line in dat:
    line = convert_line(line)
    datdict[line[0]] = line[1]
datdict

print(count_bags(datdict['shiny gold']))
